File: server-csharp/General/Reducers.cs
using System.Diagnostics.Contracts;
using System.Numerics;
using SpacetimeDB;
using System.Linq;

public static partial class Module
{
    [Reducer(ReducerKind.Init)] // The `init` parameter passed to the reducer macro indicates to SpacetimeDB that it should be called once upon database creation.
    public static void Init(ReducerContext ctx)
    {
        Log.Info($"Initializing...");

        ctx.Db.Map.Insert(new Map {Name = "Floor", Collider = FloorCollider});
        ctx.Db.Map.Insert(new Map {Name = "Ramp", Collider = RampCollider});
        ctx.Db.Map.Insert(new Map {Name = "Ramp2", Collider = Ramp2Collider});
        ctx.Db.Map.Insert(new Map {Name = "Platform", Collider = PlatformCollider});

        ctx.Db.move_all_magicians.Insert(new Move_All_Magicians_Timer
        {
            scheduled_at = new ScheduleAt.Interval(TimeSpan.FromMilliseconds(1000.0 / 60.0)),
            tick_rate = 1.0f / 60.0f
        });

        ctx.Db.handle_magician_timers_timer.Insert(new Handle_Magician_Timers_Timer
        {
            scheduled_at = new ScheduleAt.Interval(TimeSpan.FromMilliseconds(1000.0 / 60.0)),
            tick_rate = 1.0f / 60.0f
        });

        ctx.Db.gravity_magician.Insert(new Gravity_Timer_Magician
        {
            scheduled_at = new ScheduleAt.Interval(TimeSpan.FromMilliseconds(1000.0 / 60.0)),
            tick_rate = 1.0f / 60.0f,
            gravity = 20
        });

        var Match = ctx.Db.match.Insert( new Match { maxPlayers = 12, currentPlayers = 1, inProgress = false });  

        //Test Player
        ctx.Db.magician.Insert(new Magician
        {
            identity = new Identity(),
            Id = 10000,
            Name = "Test Magician",
            MatchId = Match.Id,
            Position = new DbVector3 { x = 0f, y = 0f, z = 5f},
            Rotation = new DbRotation2 { Yaw = 180, Pitch = 0 },
            Velocity = new DbVector3 { x = 0, y = 0, z = 0f},
            CorrectedVelocity = new DbVector3 { x = 0, y = 0, z = 0 },
            Collider = MagicianIdleCollider,
            CollisionEntries = [new CollisionEntry(CollisionEntryType.Map, id: 1)],
            IsColliding = true,
            KinematicInformation = new KinematicInformation(jump: false, falling: false, crouched: false, grounded: true, sprinting: false),
            State = MagicianState.Default,
            PlayerPermissionConfig =
            [
                new("CanWalk", []),
                new("CanRun", []),
                new("CanJump", []),
                new("CanAttack", []),
                new("CanCrouch", [])
            ],
            Timers = [new Timer("Attack", 0.7f, 0.7f)]
        });
    }


    [Reducer(ReducerKind.ClientConnected)]
    public static void Connect(ReducerContext ctx)
    {
        Log.Info($"{ctx.Sender} just connected.");
        var player = ctx.Db.logged_out_players.identity.Find(ctx.Sender);
        if (player != null)
        {
            ctx.Db.logged_in_players.Insert(player.Value);
            ctx.Db.logged_out_players.identity.Delete(player.Value.identity);
        }
        else
        {
            ctx.Db.logged_in_players.Insert(new Player
            {
                identity = ctx.Sender,
                Name = "Test Player"
            });
        }

        var Player = ctx.Db.logged_in_players.identity.Find(ctx.Sender) ?? throw new Exception("Player not found after insert/restore");

        var Matches = ctx.Db.match.Started.Filter(false);
        var MatchesList = Matches.ToList();
        var Match = MatchesList.Count > 0 ? MatchesList[0] : ctx.Db.match.Insert( new Match { maxPlayers = 12, currentPlayers = 0, inProgress = false });
    
        Match.currentPlayers += 1;
        ctx.Db.match.Id.Update(Match);

        ctx.Db.magician.Insert(
            new Magician
            {
                identity = Player.identity,
                Id = Player.Id,
                Name = Player.Name,
                MatchId = Match.Id,
                Position = new DbVector3 { x = 0f, y = 0f, z = 0f},
                Rotation = new DbRotation2 { Yaw = 0, Pitch = 0 },
                Velocity = new DbVector3 { x = 0, y = 0, z = 0 },
                CorrectedVelocity = new DbVector3 { x = 0, y = 0, z = 0 },
                Collider = MagicianIdleCollider, 
                CollisionEntries = [new CollisionEntry(CollisionEntryType.Map, id: 1)],
                IsColliding = true,
                KinematicInformation = new KinematicInformation(jump: false, falling: false, crouched: false, grounded: true, sprinting: false),
                State = MagicianState.Default,
                PlayerPermissionConfig =
                [
                    new("CanWalk", []),
                    new("CanRun", []),
                    new("CanJump", []),
                    new("CanAttack", []),
                    new("CanCrouch", [])
                ],
                Timers = [new Timer("Attack", 0.7f, 0.7f)]
            });
    }

    [Reducer(ReducerKind.ClientDisconnected)]
    public static void Disconnect(ReducerContext ctx)
    {
        var player = ctx.Db.logged_in_players.identity.Find(ctx.Sender) ?? throw new Exception("Player not found");
        var character = ctx.Db.magician.identity.Find(ctx.Sender) ?? throw new Exception("Magician not found");
        
        // Removes Player From Match
        var Match = ctx.Db.match.Id.Find(character.MatchId);
        if (Match != null && Match.Value.currentPlayers > 0)
        {
            var match = Match.Value;
            match.currentPlayers -= 1;
            ctx.Db.match.Id.Update(match);
        }

        ctx.Db.magician.Id.Delete(character.Id);
        
        // Removes Collision Entries Of Exited Player
        CollisionEntry CollisionEntry = new(CollisionEntryType.Magician, character.Id);
        foreach (Magician Magician in ctx.Db.magician.MatchId.Filter(character.MatchId))
        {
            if (Magician.CollisionEntries.Contains(CollisionEntry)) Magician.CollisionEntries.Remove(CollisionEntry); 
            ctx.Db.magician.Id.Update(Magician);
        }

        ctx.Db.logged_out_players.Insert(player);
        ctx.Db.logged_in_players.identity.Delete(player.identity);

        Log.Info("Disconnected");
    }
        
}

File: server-rust/spacetimedb/src/General/Reducers.rs
use std::time::Duration;
use spacetimedb::{reducer, ReducerContext, ScheduleAt, Table, TimeDuration};
use crate::*;

#[reducer(init)]
pub fn init(ctx: &ReducerContext) {
    log::info!("Initializing...");

    ctx.db.map().insert(Map {id: 0, name: "Floor".to_string(), collider: floor_collider() });
    ctx.db.map().insert(Map {id: 0, name: "Ramp".to_string(), collider: ramp_collider() });
    ctx.db.map().insert(Map {id: 0, name: "Ramp2".to_string(), collider: ramp_2_collider() });
    ctx.db.map().insert(Map {id: 0, name: "Platform".to_string(), collider: platform_collider() });
}

#[reducer(client_connected)]
pub fn connect(ctx: &ReducerContext) {
    if IsUnitTestModeEnabled(ctx) {
        return;
    }

    log::info!("{} just connected.", ctx.sender);

    let logged_out_player_option = ctx.db.logged_out_players().identity().find(ctx.sender);

    if logged_out_player_option.is_some() {
        let logged_out_player = logged_out_player_option.unwrap();
        ctx.db.logged_in_players().insert(logged_out_player);
        ctx.db.logged_out_players().identity().delete(ctx.sender);
    } 

    else {
        ctx.db.logged_in_players().insert(Player {id: 0, identity: ctx.sender, name: "Test Player".to_string() });
    }

}

#[reducer(client_disconnected)]
pub fn disconnect(ctx: &ReducerContext) {
    if IsUnitTestModeEnabled(ctx) {
        return;
    }

    let player = ctx.db.logged_in_players().identity().find(ctx.sender).expect("Player not found");

    let magician_option = ctx.db.magician().identity().find(ctx.sender);
    let respawn_timer_option = ctx.db.respawn_timers().identity().find(ctx.sender); 

    if let Some(mut magician) = magician_option {
        cleanup_on_disconnect_or_death(ctx, &mut magician);

        let game_option = ctx.db.game().id().find(magician.game_id);
        if game_option.is_some() {
            let mut game = game_option.unwrap();
            if game.current_players > 0 {
                game.current_players -= 1;
                ctx.db.game().id().update(game);
            }
        }

        ctx.db.magician().identity().delete(player.identity);
    }

    else if let Some(respawn_timer) = respawn_timer_option {
        let game_option = ctx.db.game().id().find(respawn_timer.game_id);
        if game_option.is_some() {
            let mut game = game_option.unwrap();
            if game.current_players > 0 {
                game.current_players -= 1;
                ctx.db.game().id().update(game);
            }
        }

        ctx.db.respawn_timers().scheduled_id().delete(respawn_timer.scheduled_id); 
    }
    
    ctx.db.logged_in_players().identity().delete(player.identity);
    ctx.db.logged_out_players().insert(player);
    
    log::info!("{} just disconnected.", ctx.sender);
}

#[reducer]
pub fn handle_game_end(ctx: &ReducerContext, timer: GameTimersTimer) 
{ 
    log::info!("Game Ended With Id {}", timer.game_id);
    let game_id = timer.game_id;
    cleanup_on_game_end(ctx, game_id);
    ctx.db.game_timers().scheduled_id().delete(timer.scheduled_id);
    ctx.db.game().id().delete(game_id);
}

#[reducer]
pub fn try_join_game(ctx: &ReducerContext) 
{
    let player_option = ctx.db.logged_in_players().identity().find(ctx.sender);
    if let Some(player) = player_option {
        let mut game: Game = match ctx.db.game().in_progress().filter(false).next() {
            Some(existing_game) => existing_game,
            None => {
                let created_game = ctx.db.game().insert(Game { id: 0, max_players: 12, current_players: 1, in_progress: false });
                let tick_millis: u64 = 1000 / 60;
                let tick_rate: f32 = 1.0 / 60.0;

                ctx.db.move_all_magicians().insert(MoveAllMagiciansTimer { scheduled_id: 0, scheduled_at: ScheduleAt::Interval(Duration::from_millis(tick_millis).into()), tick_rate, game_id: created_game.id });
                ctx.db.handle_magician_timers_timer().insert(HandleMagicianTimersTimer { scheduled_id: 0, scheduled_at: ScheduleAt::Interval(Duration::from_millis(tick_millis).into()), tick_rate, game_id: created_game.id });
                ctx.db.handle_magician_stateless_timers_timer().insert(HandleMagicianStatelessTimersTimer { scheduled_id: 0, scheduled_at: ScheduleAt::Interval(Duration::from_millis(tick_millis).into()), tick_rate, game_id: created_game.id });
                ctx.db.gravity_magician().insert(GravityTimerMagician { scheduled_id: 0, scheduled_at: ScheduleAt::Interval(Duration::from_millis(tick_millis).into()), tick_rate, gravity: 20.0, game_id: created_game.id });
                ctx.db.player_effects_table_timer().insert(PlayerEffectsTableTimer {scheduled_id: 0, scheduled_at: ScheduleAt::Interval(Duration::from_millis(tick_millis).into()), tick_rate, game_id: created_game.id });
                
                create_test_player(ctx, created_game.id);
                created_game
            }
        };

        game.current_players += 1;
        if game.current_players >= 2 && game.in_progress != true {
            game.in_progress = true;
            let end_time = ctx.timestamp.checked_add(TimeDuration::from_micros(600_000_000)).expect("Match End Time Timestamp Overflow"); // 10 Minutes
            let game_end_timer = GameTimersTimer {scheduled_id: 0, scheduled_at: ScheduleAt::Time(end_time), game_id: game.id};
            ctx.db.game_timers().insert(game_end_timer);
        }

        let game_id = game.id;
        ctx.db.game().id().update(game);

        let magician_config = MagicianConfig {player, game_id: game_id, position: DbVector3 { x: 0.0, y: 0.0, z: 0.0 }};
        let magician = create_magician(magician_config);
        let inserted_magician = ctx.db.magician().insert(magician);

        let invincible_effect = create_invicible_effect(5.0);
        let effects = vec![invincible_effect];
        add_effects_to_table(ctx, effects, inserted_magician.id, inserted_magician.id, game_id);   
    } 
}


#[reducer]
pub fn try_leave_game(ctx: &ReducerContext) 
{
    let player_option = ctx.db.logged_in_players().identity().find(ctx.sender);
    let magician_option = ctx.db.magician().identity().find(ctx.sender);
    let respawn_timer_option = ctx.db.respawn_timers().identity().find(ctx.sender); 

    if let Some(player) = player_option {
        if let Some(mut magician) = magician_option {
            cleanup_on_disconnect_or_death(ctx, &mut magician);

            let game_option = ctx.db.game().id().find(magician.game_id);
            if game_option.is_some() {
                let mut game = game_option.unwrap();
                if game.current_players > 0 {
                    game.current_players -= 1;
                    ctx.db.game().id().update(game);
                }
            }

            ctx.db.magician().identity().delete(player.identity);
        }

        else if let Some(respawn_timer) = respawn_timer_option {
            let game_option = ctx.db.game().id().find(respawn_timer.game_id);
            if game_option.is_some() {
                let mut game = game_option.unwrap();
                if game.current_players > 0 {
                    game.current_players -= 1;
                    ctx.db.game().id().update(game);
                }
            }

            ctx.db.respawn_timers().scheduled_id().delete(respawn_timer.scheduled_id); 
        }
    }
}

#[reducer]
pub fn handle_respawn(ctx: &ReducerContext, timer: RespawnTimersTimer) 
{ 
    let player_option = ctx.db.logged_in_players().identity().find(timer.player.identity);
    if player_option.is_none() {
        ctx.db.respawn_timers().scheduled_id().delete(timer.scheduled_id);
        return;
    }

    let player = player_option.expect("Player Existence Already Confirmed!");
    let game_option = ctx.db.game().id().find(timer.game_id);
    if game_option.is_some() {
        let magician_config = MagicianConfig {player, game_id: timer.game_id, position: DbVector3 { x: 0.0, y: 0.0, z: 0.0 }};
        let magician = create_magician(magician_config);
        let inserted_magician = ctx.db.magician().insert(magician);

        let invincible_effect = create_invicible_effect(5.0);
        let effects = vec![invincible_effect];
        add_effects_to_table(ctx, effects, inserted_magician.id, inserted_magician.id, timer.game_id);       
    }
        
    ctx.db.respawn_timers().scheduled_id().delete(timer.scheduled_id);   
}